### ISSUE TITLE
Language server doesn't report the use of undefined symbols

### ISSUE THREAD

**Issue by eldar** [DetachHead/basedpyright#513]
Hi, thank you for this excellent LSP! I'm on NeoVIM and basedpyright reports most of the issues I care about, except the use of undefined symbols like undeclared variable `x` on the screenshot:

![based_pyright](https://github.com/user-attachments/assets/ae5691a3-5705-472d-918f-322e4a9d2cd9)

This is my config where I explicitly set `reportUndefinedVariable`:

```
          lspconfig.basedpyright.setup({
            -- settings = settings("basedpyright"),
            settings = {
                basedpyright = {
                  analysis = {
                    useLibraryCodeForTypes = true,
                    diagnosticSeverityOverrides = {
                      -- https://github.com/microsoft/pyright/blob/main/docs/configuration.md
                      reportUndefinedVariable = "error",
                      reportUnusedVariable = "warning", -- or anything
                      reportUnknownParameterType = false,
                      reportUnknownArgumentType = false,
                      reportUnknownLambdaType = false,
                      reportUnknownVariableType = false,
                      reportUnknownMemberType = false,
                      reportMissingParameterType = false,
                    },
                    typeCheckingMode = "off",
                  },
                },
            },
            capabilities = capabilities,
            root_dir = require("lspconfig.util").root_pattern(
              "pyproject.toml",
              ".git"
            ),
```

I am not 100% sure if this is a bug or a missed config flag. VS Code correctly reports `reportUndefinedVariable`:

![pylance_vscode](https://github.com/user-attachments/assets/d8db9a45-ba47-4583-b7c5-675a72e32f84)

**Comment by KotlinIsland** [comment]
Does the basedpyright plugin for vscode work here?

**Comment by eldar** [comment]
It did work in VS Code and the reason is that I set the `typeCheckingMode="off"` - which overrode the value of `reportUndefinedVariable`. Looks like `diagnosticSeverityOverrides` is not a correct category for this setting, but anyway this is not a bug.

**Comment by eldar** [comment]
Closing.

**Comment by DetachHead** [comment]
`diagnosticSeverityOverrides` should be overriding whatever `typeCheckingMode` is set to, so i think there might be an issue here, unless there's something i'm missing?

**Comment by eldar** [comment]
Yes, exactly. I'm trying to set individual setting that should override `typeCheckingMode` - but that doesn't seem to work.
```
basedpyright = {
  analysis = {
    useLibraryCodeForTypes = true,
    diagnosticSeverityOverrides = {
      -- https://github.com/microsoft/pyright/blob/main/docs/configuration.md
      reportUndefinedVariable = "error",
      reportUnusedVariable = "warning", -- or anything
      reportUnknownParameterType = false,
      reportUnknownArgumentType = false,
      reportUnknownLambdaType = false,
      reportUnknownVariableType = false,
      reportUnknownMemberType = false,
      reportMissingParameterType = false,
    },
    typeCheckingMode = "off",
  },
},
```
Is this correct structure?

**Comment by DetachHead** [comment]
so `diagnosticSeverityOverrides` work on all the other `typeCheckingMode`s but not `"off"`? if that's the case then it's definitely a bug, in which case we should re-open this issue.

i have a project that's set up this way using `pyproject.toml` and it seems to work, so maybe this issue only occurs in the language server?

**Comment by eldar** [comment]
`diagnosticSeverityOverrides` doesn't work for me either way, regardless of the value of `typeCheckingMode` - which I also set as `basic` or `standard`. Perhaps this is a problem with how `nvim-lspconfig` is setting up basedpyright, but then again typeCheckingMode does work. I can check whether setting these in `pyproject.toml` works (I presume it would) - the one I posted here done in Lua's configuration of NeoVIM.

**Comment by DetachHead** [comment]
do you have a `tool.basedpyright` or `tool.pyright` section in your `pyproject.toml`? if so, does deleting it make the `diagnosticSeverityOverrides` in your lsp config work?

**Comment by eldar** [comment]
I tried `pyproject.toml` and it works:

```toml
[tool.basedpyright]                                                                                                                                                                                                        
typeCheckingMode = "basic"                                                                                                                                                                                               
reportDuplicateImport = "warning"
```

However setting the same from LSP config in NeoVIM doesn't - and I tried placing `reportDuplicateImport` both in `analysis` section or in `diagnosticSeverityOverrides` - in both cases it is ignored:

```lua
basedpyright = function()                                                                                                                                                                          
   lspconfig.basedpyright.setup({                                                                                                                                                                  
     -- settings = settings("basedpyright"),                                                                                                                                                       
     settings = {                                                                                                                                                                                  
       basedpyright = {                                                                                                                                                                            
         analysis = {                                                                                                                                                                              
           useLibraryCodeForTypes = true,                                                                                                                                                          
           typeCheckingMode = "basic", -- off, basic, standard, strict, all                                                                                                                        
           reportDuplicateImport = "warning",                                                                                                                                                      
           diagnosticSeverityOverrides = {                                                                                                                                                         
             reportDuplicateImport = "warning",                                                                                                                                                    
           },                                                                                                                                                                                      
         },                                                                                                                                                                                        
       },                                                                                                                                                                                          
     },                                                                                                                                                                                            
     capabilities = capabilities,                                                                                                                                                                  
     root_dir = require("lspconfig.util").root_pattern(                                                                                                                                            
       "pyproject.toml",                                                                                                                                                                           
       ".git"                                                                                                                                                                                      
     ),                                                                                                                                                                                            
   })  
```

**Comment by DetachHead** [comment]
i can't seem to reproduce this unfortunately. do you think you could create a minimal reproduction of the issue, similar to [the one in this comment](https://github.com/DetachHead/basedpyright/issues/492#issuecomment-2223736189)?

**Comment by eldar** [comment]
Hey! I managed to narrow down the problem and prepared a test case. Here's the neovim config and corresponding python test case and `pyproject.toml`:

```sh
# download the NeoVIM config with Pyright
export NVIM_APPNAME=basedpyright-bugreport
mkdir ~/.config/$NVIM_APPNAME
cd ~/.config/$NVIM_APPNAME
wget https://gist.githubusercontent.com/eldar/8a79dd4d84b63ed277667401f6cd3e81/raw/397deeb5f8f4a1730ad0d1358cbce99cd5998e5b/init.lua

# then clone the python test case somewhere
cd $HOME
git clone https://github.com/eldar/basedPyright-issue-513.git
cd basedPyright-issue-513
nvim test.py
```

Now, to the core of the issue. `init.lua` configures basedPyright which works by itself:
```lua
typeCheckingMode = "basic", -- off, basic, standard, strict, all
diagnosticSeverityOverrides = {
  reportDuplicateImport = "warning",
},
```

However in my testcase I included `pyproject.toml` with a single line:
```
[tool.basedpyright]
```

Enabling basedPyright config in `pyproject.toml` makes the settings provided in `init.lua` irrelevant, even though there is no `typeCheckingMode` or `diagnosticSeverityOverrides` in the TOML file. You can verify that by commenting out `[tool.basedpyright]` line in `pyproject.toml` which will enable detection of duplicate imports as specified in `init.lua`.

The expected behavior here is that individual settings provided in `pyproject.toml` should take priority over global settings in the editor, but merely including a section in the project specific file should not invalidate the entire config. Thanks for looking into this!

**Comment by DetachHead** [comment]
i think this is the same issue as https://github.com/microsoft/pyright/issues/836

tbh this is one of the few times where i agree with eric, specifically this comment https://github.com/microsoft/pyright/issues/836#issuecomment-659808974 where he says if this was supported, it would result in users seeing different errors to what they'd see in their CI.

this has been raised before in [the basedtyping discord](https://discord.gg/7y9upqPrk2), and we came to an agreement that the current behavior is acceptable, but there should be some sort of warning visible to the user that this occurs (#67) instead of just silently ignoring the settings.

**Comment by eldar** [comment]
Hey, thanks for tracking it down to the upstream issue!

I would still argue that the original behavior is not particularly thought through. More specifically, it basically renders the default settings in the editor/IDE useless. I added basedpyright section to `pyproject.toml` because I wanted to add workspace-specific exclude paths which seems like a common thing to do. Also, the same usecase (using project specific configuration for exclude paths) was reported by the author of the issue in the upstream Pyright repository. At the same time, I feel that configuring LSP diagnostics is something that most people do quire rarely and it's easier if it is a part of the editor configuration.

The workaround is simple enough though which is just to copy around corresponding pyright options between different projects, so this is not really a blocker.

**Comment by DetachHead** [comment]
fair enough, several people have raised this both here and upstream so i think it's reasonable to support it

**Comment by rbhanot4739** [comment]
> i do think this functionality should be opt-in though, since i personally would never want any global config to override project-specific config.

@DetachHead  But this can be argued the other way around, as well as the mere presence of `pyrightconfig.json` or `[based]pyrightconfig.json` sections in `pyproject.toml` would simply wipe out all global defaults that a user might not want to repeat in every project's config file. 

As per me the ideal sane behaviour should be to merge the global config with local project config files (`pyproject.toml` and `[based]pyrightconfig.json`) take precedence over it. This is how VsCode also merges the user settings with folder/workspace settings.

> come to think of it, that kinda sounds like the "extends" option which already exists. does that cover this use case?

Is this supported yet ?

**Comment by DetachHead** [comment]
> As per me(unless I am missing something) the ideal sane behaviour should be to merge the global config with local project config files (`pyproject.toml` and `[based]pyrightconfig.json`) take precedence over it. This is how VsCode also merges the user settings with folder/workspace settings.

this is how pyright used to behave until recently (version 1.1.376 i believe). the reasoning behind this change, which i agree with, is that when a project has defined its own pyright config, it should not be overridden by individual users as all the code inside that project was written with a specific set of rules in mind, and likely has CI configured to run pyright with the exact same config.

so in order to prevent users from seeing different errors in their IDE to their CI (something i've encountered many times and it always annoyed me), or from others working on the same project, this behavior is necessary. otherwise it's too easy to become desensitized to the red underline because it's "only an error locally" (kinda related: https://github.com/DetachHead/basedpyright/issues/603#issuecomment-2303373090)

> > come to think of it, that kinda sounds like the "extends" option which already exists. does that cover this use case?
> 
> Is this supported yet ?

yes, see [here](https://docs.basedpyright.com/#/configuration?id=extends):

> - **extends** [path, optional]: Path to another `.json` or `.toml` file that is used as a “base configuration”, allowing this configuration to inherit configuration settings. Top-level keys within this configuration overwrite top-level keys in the base configuration. Multiple levels of inheritance are supported. Relative paths specified in a configuration file are resolved relative to the location of that configuration file.

# TL;DR - my stance on this issue, for anyone stumbling across this as this thread is pretty long now

i believe the correct solution is to copy your local config into the project config as mentioned above in https://github.com/DetachHead/basedpyright/issues/513#issuecomment-2278058048. i don't see this as a "workaround" but rather the correct way to configure your project. doing so also helps other contributors because it gives them the project's expected config by default without them having to change their local settings too. it also ensures that the errors reported by the CLI and IDE are always consistent.

i think the only problem i need to address is the fact that it's not obvious to the user when this is happening. an language server notification saying something like "This project has basedpyright config in `pyproject.toml` which will override your user settings." along with a link to some documentation explaining these points should be sufficient

**Comment by rbhanot4739** [comment]
@DetachHead Thank you for the comprehensive response, makes perfect sense.

**Comment by munkybutt** [comment]
thank you for this thread! I have been going around in circles trying to figure out why non of my settings have been working!

**Comment by heygarrett** [comment]
I think I'm having the opposite problem, though correct me if I'm not fully understanding the intended functionality.

If I've disabled a rule in my language server config in Neovim:

```lua
require("lspconfig")["basedpyright"].setup({
	settings = {
		basedpyright = {
			analysis = {
				diagnosticSeverityOverrides = {
					reportImplicitStringConcatenation = false,
				},
			},
		},
	},
})
```

But the `pyproject.toml` in a project has _not_ disabled that rule:

```toml
[tool.pyright]
typeCheckingMode = "strict"
```

I would expect that rule to be enabled within the project. But it's still disabled for me. Is this behavior correct or incorrect? And depending, should I open a new issue?

**Comment by DetachHead** [comment]
`"strict"` mode does not enable every rule. `reportImplicitStringConcatenation` is only enabled by default in `"recommended"` and `"all"` rulesets. see [here](https://docs.basedpyright.com/latest/configuration/config-files/#diagnostic-settings-defaults) for the defaults

**Comment by heygarrett** [comment]
Ah, my mistake. Sorry for the noise!

**Comment by j-palms** [comment]
> Something I don't understand is: There are two kinds of settings:

* ["Language Server Settings" ](https://docs.basedpyright.com/v1.22.1/configuration/language-server-settings/#based-settings) configured in the IDE settings (like `basedpyright.analysis.diagnosticMode = "workspace"`)
 
* ["configuration options"](https://docs.basedpyright.com/v1.22.1/configuration/config-files/#sample-pyprojecttoml-file) specified in a JSON-formatted text configuration (pyrightconfig.json/pyproject.toml)
> 
> 
> But at the same time, the Language Server Settings are ignored if there is a `pyrightconfig.json` file? Shouldn't it then be possible to set things like `basedpyright.analysis.diagnosticMode = "workspace"` in the `pyrightconfig.json`?

I had the above question and was redirected here, but to me it still makes no sense that the **full** IDE LSP configuration is ignored in case of the existence of a project configuration.
First of all, I completely understand the reason why the project configuration should overwrite the global config. But since it is **impossible** to set something like the `diagnosticMode` in the project config (which also makes sense because this purely personal "ui" preference and is not related to the project itself), it should be still possible to set it in the global config. 
This was also fixed in pylance: https://github.com/microsoft/pylance-release/issues/6314

**Comment by DetachHead** [comment]
my bad, i didn't read your whole comment in https://github.com/DetachHead/basedpyright/issues/894#issuecomment-2520846030 before directing you to this issue. `basedpyright.analysis.diagnosticMode` should still be configureable in the language server settings regardless of whether you have a config file in your project.

this behavior only effects config that actually *can* be overwritten in the `pyproject.toml` / `pyrightconfig.json` files, so since that setting cannot be set in the config files, i'm not sure why it's not working for you.

could you provide some more info such as which editor are you using, which version of basedpyright you're using, and what your language server config looks like
